data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/ScanObjectWorker.java

```diff
@@ -36,118 +37,126 @@ public class ScanObjectWorker implements Runnable{
-                .filter(path -> !path.endsWith("/"))
-                .filter(includeKeyPath -> excludeKeyPaths.stream()
-                        .noneMatch(excludeItem -> includeKeyPath.endsWith(excludeItem))).collect(Collectors.toList());
-    }
+    private void startProcessingObject(final int waitTimeMillis) {
+        final Optional<SourcePartition<S3SourceProgressState>> objectToProcess = sourceCoordinator.getNextPartition(this::provideFilteredBucketKeyPartitions);
 
-    private void processS3ObjectKeys(final S3ObjectReference s3ObjectReference,
-                                     final S3ObjectHandler s3ObjectHandler,
-                                     final ScanOptions scanOptions){
-        final S3ObjectDetails s3ObjDetails = getS3ObjectDetails(s3ObjectReference);
-        final boolean isKeyMatchedBetweenTimeRange = isKeyMatchedBetweenTimeRange(s3ObjDetails.getS3ObjectLastModifiedTimestamp(),
-                scanOptions.getUseStartDateTime(),
-                scanOptions.getUseEndDateTime());
-        if(isKeyMatchedBetweenTimeRange && (isKeyNotProcessedByS3Scan(s3ObjDetails))){
-            updateKeyProcessedByS3Scan(s3ObjDetails);
-            try{
-                s3ObjectHandler.parseS3Object(s3ObjectReference,null);
-            }catch (IOException ex){
-                deleteKeyProcessedByS3Scan(s3ObjDetails);
-                LOG.error("Error while process the parseS3Object. ",ex);
+        if (objectToProcess.isEmpty()) {
+            try {
+                Thread.sleep(waitTimeMillis);
+            } catch (InterruptedException e) {
+                e.printStackTrace();
             }
+            return;
+        }
+
+        final String bucket = objectToProcess.get().getPartitionKey().split("\\|")[0];
+        final String objectKey = objectToProcess.get().getPartitionKey().split("\\|")[1];
+
+        try {
+            processS3Object(S3ObjectReference.bucketAndKey(bucket, objectKey).build());
+            sourceCoordinator.completePartition(objectToProcess.get().getPartitionKey());
+        } catch (final PartitionNotOwnedException | PartitionNotFoundException | PartitionUpdateException e) {
+            LOG.warn("S3 scan object worker received an exception from the source coordinator. There is a potential for duplicate data from {}, giving up partition and getting next partition: {}", objectKey, e.getMessage());
+            sourceCoordinator.giveUpPartitions();
         }
     }
-    /**
-     * Method will identify already processed key.
-     * @return boolean
-     */
-    private boolean isKeyNotProcessedByS3Scan(final S3ObjectDetails s3ObjectDetails) {
-        return stateSaveMap.get(s3ObjectDetails.getBucket()+s3ObjectDetails.getKey()) == null;
+
+    private List<PartitionIdentifier> provideFilteredBucketKeyPartitions() {
+        final List<PartitionIdentifier> objectsToProcess = new ArrayList<>();
+
+        for (final ScanOptions scanOptions : scanOptionsBuilderList) {
```

Comment:
What keeps this from scanning the same items on every call?

Reply:
This is being called within the `run()` method inside a continual loop. So won't it list objects each time this is called?

Reply:
This is not being called in the run method. This is being passed to the source coordinator, and the source coordinator will only run the supplier when there are no available partitions (https://github.com/opensearch-project/data-prepper/blob/abce720ecef25f23a2109d9316c63206c3cb48dd/data-prepper-core/src/main/java/org/opensearch/dataprepper/sourcecoordination/LeaseBasedSourceCoordinator.java#L87)

Reply:
Thanks for pointing that out. This makes sense now. So I think this can work for now.

I still think there is something about the behavior here which can hit issues at large scale. If you have 1000s of objects in a bucket, each time you run out of partitions you must scan all objects all over again. Most use-cases can probably assume some mechanism for the order of growth. In S3, this can be the key order, for the OpenSearch source this is the query sort.

This call `partitionsCreatorSupplier.get()` does not track progress. I think we might want to consider making this a function instead.

The input into that function could be some reference as to what the last object scanned was. But, this should go into another issue.